File: android/app/src/main/kotlin/com/anapaya/chat/app/ui/chat/ChatRows.kt
// Copyright 2026 Anapaya Systems

package com.anapaya.chat.app.ui.chat

import com.anapaya.chat.client.model.Message
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

/** A row of the conversation: something somebody said, or the day it was said on. */
internal sealed interface ChatRow {
    val key: String

    data class Day(val label: String) : ChatRow {
        override val key: String get() = "day $label"
    }

    /** One message, and where it sits in its author's run. */
    data class Said(
        val message: Message,
        val clock: String,
        val mine: Boolean,
        val opensGroup: Boolean,
        val closesGroup: Boolean,
        /** A name under this message carries the gap, so the row leaves none of its own. */
        val nameFollows: Boolean,
    ) : ChatRow {
        override val key: String get() = "seq ${message.seq}"
    }
}

/** The clock and the date. [Locale.ROOT] keeps it at 24 hours, as the terminal client prints it. */
internal class Clocks(private val now: () -> Long = System::currentTimeMillis) {
    private val time = SimpleDateFormat("HH:mm", Locale.ROOT)
    private val date = SimpleDateFormat("d MMM", Locale.ROOT)
    private val on = Calendar.getInstance()

    fun clock(at: Long): String = time.format(Date(at))

    /** Which day `at` falls on locally, as one comparable number. */
    fun dayOf(at: Long): Long {
        on.timeInMillis = at
        return on.get(Calendar.YEAR) * 1000L + on.get(Calendar.DAY_OF_YEAR)
    }

    /** What to write above the day's first message. */
    fun label(at: Long): String {
        val today = dayOf(now())
        return when (dayOf(at)) {
            today -> "TODAY"
            today - 1 -> "YESTERDAY"
            else -> date.format(Date(at)).uppercase(Locale.ROOT)
        }
    }
}

/** The messages as rows, oldest first: the only order a change of day or of author is visible in. */
internal fun chatRows(messages: List<Message>, me: String?, clocks: Clocks): List<ChatRow> {
    val rows = ArrayList<ChatRow>(messages.size + 2)
    var day: Long? = null

    messages.forEachIndexed { index, message ->
        val on = clocks.dayOf(message.postedAt)
        val newDay = on != day
        if (newDay) {
            day = on
            rows.add(ChatRow.Day(clocks.label(message.postedAt)))
        }

        val before = messages.getOrNull(index - 1)
        val after = messages.getOrNull(index + 1)
        // A day separator breaks a run, so the name is written again under it.
        val opens = newDay || before?.username != message.username
        val closes = after == null ||
            after.username != message.username ||
            clocks.dayOf(after.postedAt) != on

        // A day separator carries its own space, so only a name on the same day counts.
        val nameFollows = closes &&
            after != null &&
            after.username != me &&
            clocks.dayOf(after.postedAt) == on

        rows.add(
            ChatRow.Said(
                message = message,
                clock = clocks.clock(message.postedAt),
                mine = message.username == me,
                opensGroup = opens,
                closesGroup = closes,
                nameFollows = nameFollows,
            ),
        )
    }

    return rows
}

File: ios/App/ChatRows.swift
// Copyright 2026 Anapaya Systems

import ChatClient
import Foundation

/// A row of the conversation: something somebody said, or the day it was said on.
enum ChatRow: Identifiable {
    case day(String, id: String)
    case said(Said)

    var id: String {
        switch self {
        case .day(_, let id): return id
        case .said(let said): return "seq \(said.message.seq)"
        }
    }

    /// One message, and where it sits in its author's run.
    struct Said {
        let message: Message
        let clock: String
        let mine: Bool
        let opensGroup: Bool
        let closesGroup: Bool
        /// A name under this message carries the gap, so the row leaves none of its own.
        let nameFollows: Bool
    }
}

/// The clock and the date. A fixed locale keeps it at 24 hours, as the terminal client prints it.
struct Clocks {
    private let time: DateFormatter
    private let date: DateFormatter
    private let calendar = Calendar.current
    private let now: () -> Date

    init(now: @escaping () -> Date = Date.init) {
        self.now = now
        time = DateFormatter()
        time.locale = Locale(identifier: "en_US_POSIX")
        time.dateFormat = "HH:mm"
        date = DateFormatter()
        date.locale = Locale(identifier: "en_US_POSIX")
        date.dateFormat = "d MMM"
    }

    func clock(_ at: Int64) -> String { time.string(from: stamp(at)) }

    /// Which day `at` falls on locally, as one comparable number.
    func dayOf(_ at: Int64) -> Int {
        Int(calendar.startOfDay(for: stamp(at)).timeIntervalSince1970)
    }

    /// What to write above the day's first message.
    func label(_ at: Int64) -> String {
        let today = calendar.startOfDay(for: now())
        let day = calendar.startOfDay(for: stamp(at))

        if day == today { return "TODAY" }
        if let yesterday = calendar.date(byAdding: .day, value: -1, to: today), day == yesterday {
            return "YESTERDAY"
        }
        return date.string(from: day).uppercased()
    }

    private func stamp(_ at: Int64) -> Date {
        Date(timeIntervalSince1970: TimeInterval(at) / 1000)
    }
}

/// The messages as rows, oldest first: the only order a change of day or of author is visible in.
func chatRows(messages: [Message], me: String?, clocks: Clocks) -> [ChatRow] {
    var rows: [ChatRow] = []
    var day: Int?

    for (index, message) in messages.enumerated() {
        let on = clocks.dayOf(message.postedAt)
        let newDay = on != day
        if newDay {
            day = on
            rows.append(.day(clocks.label(message.postedAt), id: "day \(on)"))
        }

        let before = index > 0 ? messages[index - 1] : nil
        let after = index + 1 < messages.count ? messages[index + 1] : nil
        // A day separator breaks a run, so the name is written again under it.
        let opens = newDay || before?.username != message.username
        let closes = after == nil
            || after?.username != message.username
            || clocks.dayOf(after!.postedAt) != on

        // A day separator carries its own space, so only a name on the same day counts.
        let nameFollows = closes
            && after != nil
            && after!.username != me
            && clocks.dayOf(after!.postedAt) == on

        rows.append(.said(ChatRow.Said(
            message: message,
            clock: clocks.clock(message.postedAt),
            mine: message.username == me,
            opensGroup: opens,
            closesGroup: closes,
            nameFollows: nameFollows)))
    }

    return rows
}
